# Exercise 5

Related Notes:
- Chapter_05_File_IO

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/njc-cz2-2021/Materials/blob/main/Exercises/Exercise_05_File_IO.ipynb) 

## Exercise 5.1

Write following list `s` into a text file `q1.txt` with each item occupying 1 line.

>``` python
>s = [chr(i) + '.txt' for i in range(ord('A'), ord('F'))]
>```

Use following command in Jupyter Notebook to open `q1.txt` to verify the content in the file.

>`!notepad q1.txt`

In [5]:
s = [chr(i) + '.txt' for i in range(ord('A'), ord('F'))]

with open('./output/q1.txt', 'w') as f:
    f.writelines([i + '\n' for i in s])


## Exercise 5.2

For each line of text in `q1.txt`, generate an empty text file and name the file using that line of text. If file already exists, remove it before generating a new file. 

Use the following code to check if a file exists.

>``` python
>import os
>if os.path.exists(file_path): pass
>```

Use following command in Jupyter Notebook to confirm that you have generated all files.

>``` bash
>!ls *.txt
>```

In [8]:
import os

with open('./output/q1.txt') as f:
    files = f.readlines()

for f in files:
    f = f.rstrip()
    try:
        os.remove('./output/' + f)
    except:
        pass
    with open('./output/' + f, 'w') as alphabet_file:
        alphabet_file.write(f)


        
    



## Exercise 5.3

Write code to generate all the positive integers up to $n$ that are perfect squares and save them in a text file called `PERFECT_SQUARES.TXT`. Note that each line within the file should contain one integer, and the last character within the file should be an end line character.

In [ ]:
from math import sqrt

n = 10000
perfect_squares = [i for i in range(n) if sqrt(n) % 1 == 0]



## Exercise 5.4

Write code to generate all the positive integers from $n + 1$ to $n + m$, and then append them to the file used in Exercise 5.3. (Note, do not overwrite the perfect squares from Exercise 5.3) 

In [ ]:
#YOUR_CODE_HERE

## Exercise 5.5

A log of a certain boss fight from a certain game is stored inside the file `log_ex8.csv` under the folder `resources`. The csv file contains the following header.

- `'Parse%'`  refers to the performance of the character relative to all other players in the world,
- `'Name'` refers to the name of the character in the game,
- `'Amount'`  refers to the amount of damage
- `'iLvL'`  refers to the quality of the equipment the character is using,
- `'iLvL%'` refers to the performance of the player relative to players with similar quality of equipment,
- `'Active Time'` refers to the percentage of time spend by the character doing damage during the fight.
- `'DPS'` refers to the amount of damage done per second in the fight.

Your task is to:

- Store the contents of the csv file into a list `records` where each element in the list is itself a list where the elements correspond to the `Parse%`,`Name`,`Amount`,`iLvL`,`iLvL%`,`Active Time`,`DPS` values for each line in the csv file.
- Create an empty list called `player_name`.
- iterate over each element of `records` appending the player names to `player_name`.
- Print each name in `player_name` in alphabetical order.

In [31]:
import csv
with open('./resources/log_ex8.csv') as f:
    header, *reader = csv.reader(f)
    records = [row for row in reader]
player_name = sorted([row[1] for row in records])
print(player_name)



['Angers', 'Anothervc', 'Avey', 'Biraprie', 'Chrisfumi', 'Chrissinger', 'Darkietank', 'Dewgie', 'Ericlock', 'Exolaris', 'Fyrebird', 'Gamheals', 'Gdook', 'Haloxd', 'Jari', 'Keipra', 'Kieu', 'Knoxvegas', 'Kolhito', 'Lymos', 'Makotoase', 'Masachi', 'Masakins', 'Nómi', 'Ovy', 'Paeldryth', 'Pappaslop', 'Plastral', 'Revtide', 'Spooknspell', 'Stormblast', 'Synesthetix', 'Wôes']


## Exercise 5.6

Download the file: https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data

Write code to read and store the contents of this comma separated file within a 2D List, $L$, such that $L[i]$ corresponds to the sample entry on the $i$-th line of the file, and $L[i][j]$ corresponds to the $j$-th value of the $i$-th entry.
Print the contents of the 2D List to verify that you have stored the contents correctly.


In [32]:
import csv

with open('./resources/IRIS.data') as f:
    L = [row for row in csv.reader(f)]
print(L)

[['5.1', '3.5', '1.4', '0.2', 'Iris-setosa'], ['4.9', '3.0', '1.4', '0.2', 'Iris-setosa'], ['4.7', '3.2', '1.3', '0.2', 'Iris-setosa'], ['4.6', '3.1', '1.5', '0.2', 'Iris-setosa'], ['5.0', '3.6', '1.4', '0.2', 'Iris-setosa'], ['5.4', '3.9', '1.7', '0.4', 'Iris-setosa'], ['4.6', '3.4', '1.4', '0.3', 'Iris-setosa'], ['5.0', '3.4', '1.5', '0.2', 'Iris-setosa'], ['4.4', '2.9', '1.4', '0.2', 'Iris-setosa'], ['4.9', '3.1', '1.5', '0.1', 'Iris-setosa'], ['5.4', '3.7', '1.5', '0.2', 'Iris-setosa'], ['4.8', '3.4', '1.6', '0.2', 'Iris-setosa'], ['4.8', '3.0', '1.4', '0.1', 'Iris-setosa'], ['4.3', '3.0', '1.1', '0.1', 'Iris-setosa'], ['5.8', '4.0', '1.2', '0.2', 'Iris-setosa'], ['5.7', '4.4', '1.5', '0.4', 'Iris-setosa'], ['5.4', '3.9', '1.3', '0.4', 'Iris-setosa'], ['5.1', '3.5', '1.4', '0.3', 'Iris-setosa'], ['5.7', '3.8', '1.7', '0.3', 'Iris-setosa'], ['5.1', '3.8', '1.5', '0.3', 'Iris-setosa'], ['5.4', '3.4', '1.7', '0.2', 'Iris-setosa'], ['5.1', '3.7', '1.5', '0.4', 'Iris-setosa'], ['4.6', '

## Exercise 5.7

Each line in the file at: http://itu.dk/people/pagh/sad12/primes1000000.txt represents a prime number below 1,000,000.
 
Write a function that uses the above file to test if a given positive integer below 1,000,000 is a prime number. If the given numbers exceeds 1,000,000, then utilise a separate function to calculate if it is a prime. 

In [ ]:
from math import sqrt, ceil

def get_factor(n):
    with open('./resources/primes1000000.txt') as f:
        prime = 0
        while prime ** 2 <= n and prime < 999983:
            prime = int(f.readline())
            if n % prime == 0:
                return (prime)
        return (None)

def get_factor_large(n):
    for i in range(999996, ceil(sqrt(n))+6, 6):
        if n % (i-1) == 0:
            return (i-1)
        if n % (i+1) == 0:
            return (i+1)
    return (None)
        

n = int(input("Input positive integer:")) #54192375525768481 (17d square no.) takes ~18s to compute
prime_factor = get_factor(n)
if prime_factor != None:
    print('no boohoo ',prime_factor,' is factor')
elif n < 1000000000000:
    print('yay prime')
else:
    prime_factor = get_factor_large(n)
    if prime_factor != None:
        print('no boohoo ',prime_factor,' is factor')
    else:
        print('yay prime')

no boohoo  232792559  is factor


## Exercise 5.8

A mathematical equation $y=f(x)$ can be visualized on a cartesian plane as the set of all the points $(x,y)$ where $y=f(x)$. Write a code that 

1. generates 100 points that satisfy the equation $y=x^2$ for $-10\leq x\leq 10$ (for simplicity, take the uniform spacing for the $x$-values of the points),and
2. saves them into a text file named `ex8_8.txt`. Each line in the file is of the form $(x,y)$, 
3. load the points from `ex8_8.txt` and using the module `matplotlib`, show the plot.


In [ ]:
#YOUR_CODE_HERE

## Exercise 5.9

A text file, `NOI_24_RESULT.txt` contains the results of the medallist recipients for the 24th National Olympiad in Informatics.

Write a code to:

- read the data from the text file,
- store the school name and medal counts in a Python dictionary `medal_count` where the school name is the key and the associated value is a tuple of the form `(GOLD, SILVER, BRONZE)`,
- print out the top 5 school with the highest number of medals won, regardless of the medal type. 

In [ ]:
#YOUR_CODE_HERE

## Exercise 5.10

A text file, `MARKET_PRICES.TXT`, contains the in-game prices for 6 items, Widowbloom, Rising Glory, Marrowroot, Vigil's Torch, Nighshade and Spectral Flask of Power from 1 December 2020 to 3 March 2021. Each line contains tab-delimited data that shows the date and the prices (in copper coins) of the 6 items.

Each line is in the format:

>``` text
>YYYY-MM-DD\tWIDOWBLOOM\tRISING_GLORY\t\MARROWROOT\tVIGILS_TORCH\tNIGHTSHADE\tSPECTRAL_FLASK_OF_POWER\n
>```

- The date is of the form `YYYY-MM-DD`, for example, `2020-12-03` is 3rd December, 2020.
- WIDOWBLOOM, RISING_GLORY, MARROWROOT, VIGILS_TORCH, NIGHTSHADE, SPECTRAL_FLASK_OF_POWER are all integers.
- `\t` represents the tab character
- `\n` represents the newline character

The text file is stored in ascending order of date and time.

### Task 1 

Write a program to:

- read the price data from the text file,
- for each of the 6 items:
    - find the highest price and print this value,
    - find the lowest price and print this value.

### Task 2

Suppose you want to play the market by buying and selling the 6 items above, one way to decide on items to invest in is by using the mean and variance of the daily rate of returns of the items. The latter is used to quantify the risk of holding such item.

For an item $j$, 
- the **daily rate of return** of day $i$ to $i+1$, denoted as $r_i$ is defined to be $$r_i=\frac{p_{i+1}-p_{i}}{p_i},$$ where $p_{i+1}$ and $p_i$ are prices of the item on day $i+1$ and day $i$ respectively,
- the mean of the daily rate of returns, denoted as $\overline{r}$, is just $\frac{\sum r_i}{n}$, where $n$ is the number of daily rate of returns recorded,
- the variance of the daily rate of returns is $$\frac{\sum{(r_i-\overline{r})^2}}{n},$$ where $\overline{r}$ is the mean of the daily rate of returns and again $n$ is the number of daily returns recorded.

Write a program to:

- for each of the 6 items:
    - find the mean of the daily rate of return,
    - find the variance of the daily rate of return.

In [ ]:
#YOUR_CODE_HERE